src/Interpreters/Context.cpp

```diff
@@ -1393,62 +1395,128 @@ StoragePtr Context::executeTableFunction(const ASTPtr & table_expression, const
+            auto virtual_column_names = table_function_ptr->getVirtualsToCheckBeforeUsingStructureHint();
+            bool asterisk = false;
+            const auto & expression_list = select_query_hint->select()->as<ASTExpressionList>()->children;
+            const auto * expression = expression_list.begin();
+
+            for (; expression != expression_list.end() && insert_column != insert_structure_end; ++expression)
+            {
+                if (auto * identifier = (*expression)->as<ASTIdentifier>())
+                {
+                    if (!virtual_column_names.contains(identifier->name()))
                     {
-                        have_asterisk = true;
+                        if (asterisk)
+                        {
+                            if (use_structure_from_insertion_table_in_table_functions == 1)
+                                throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Asterisk cannot be mixed with column list in INSERT SELECT query.");
+
+                            use_columns_from_insert_query = false;
+                            break;
+                        }
+
+                        structure_hint.add({ identifier->name(), insert_column->type });
                     }
-                    else if (!expression->as<ASTLiteral>())
+
+                    if (asterisk)
+                        --insert_structure_end;
+                    else
+                        ++insert_column;
+                }
+                else if ((*expression)->as<ASTAsterisk>())
+                {
+                    if (asterisk)
                     {
+                        if (use_structure_from_insertion_table_in_table_functions == 1)
+                            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Only one asterisk can be used in INSERT SELECT query.");
+
+                        use_columns_from_insert_query = false;
+                        break;
+                    }
+                    if (!structure_hint.empty())
+                    {
+                        if (use_structure_from_insertion_table_in_table_functions == 1)
+                            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Asterisk cannot be mixed with column list in INSERT SELECT query.");
+
                         use_columns_from_insert_query = false;
                         break;
                     }
-                }
 
-                /// Check that all identifiers are column names from insertion table and not virtual column names from storage.
-                for (const auto & column_name : columns_names)
+                    asterisk = true;
+                }
+                else if (auto * function = (*expression)->as<ASTFunction>())
                 {
-                    if (!structure_hint.has(column_name) || virtual_column_names.contains(column_name))
+                    if (use_structure_from_insertion_table_in_table_functions == 2 && find_identifier(function))
```

Comment:
I see that in case when `use_structure_from_insertion_table_in_table_functions=1` and `find_identifier(function)=true` we just skip this expression. Maybe better to throw an exception? Or it's ok that we will have an exception `missing column` later?

Reply:
There can be a case when having identifier in function will work - for example `select a+1,a`, so I think it's better to left this opportunity for optimization.